module/zfs/zfs_znode.c

```diff
@@ -1060,34 +1065,30 @@ zfs_zget(zfs_sb_t *zsb, uint64_t obj_num, znode_t **zpp)
 
 		mutex_enter(&zp->z_lock);
 		ASSERT3U(zp->z_id, ==, obj_num);
-		if (zp->z_unlinked) {
-			err = SET_ERROR(ENOENT);
-		} else {
```

Comment:
Removing this check to allow objects in the process of being unlinked to be returned is a significant change.  However, it turns out this is possible in the existing code since there's nothing preventing `z_unlinked` from being set after the `z_lock` mutex is dropped but before `zfs_zget()` returns.  And in fact this behavior is desirable since until the last reference on the inode is dropped we need to allow normal operations to work.